torch/csrc/jit/codegen/cuda/lower_index_hoist.cpp

```diff
@@ -12,419 +12,275 @@ namespace cuda {
+    // of a tensor, this approach will fail. For this case we just give up
+    // hoisting by returning a nullptr.
+    return nullptr;
   }
 
-  if (used_loops_.size() != other.used_loops_.size()) {
-    return false;
+  auto def = index->definition();
+  if (def == nullptr || index->isConstScalar()) {
+    return index;
   }
 
-  // Check if both CommonIndexKeys use the same loops. If not, it's
-  // still valid to share the same hoisted index as long as: 1) each
-  // loop pair is mapped with the CA index map, and 2) they are not
-  // instantiated as actual loops.
-  for (const auto i : c10::irange(used_loops_.size())) {
-    auto lhs_loop = used_loops_.at(i);
-    auto rhs_loop = other.used_loops_.at(i);
-    if (lhs_loop == rhs_loop) {
-      continue;
-    }
-    if (gpu_lower->caMap()->areMapped(
-            lhs_loop->iter_domain(),
-            rhs_loop->iter_domain(),
-            IdMappingMode::EXACT) &&
-        lhs_loop->isTrivial() && rhs_loop->isTrivial()) {
-      continue;
-    }
-    return false;
+  std::unordered_set<Val*> my_dependency =
+      getLoopVariableDependency(index, loops);
+  auto my_level = findLowestLevelWithSatisfiedDependency(loops, my_dependency);
+
+  // Check if `index` is already computed. If yes, just reuse it and return.
+  auto existing_subexpr = eliminateCommonSubexpression(index, loops, my_level);
+  if (existing_subexpr != nullptr) {
+    return existing_subexpr;
   }
 
-  for (const auto i : c10::irange(loop_index_vals_.size())) {
-    auto lhs_index = loop_index_vals_.at(i);
-    auto rhs_index = other.loop_index_vals_.at(i);
-    if (lhs_index == rhs_index) {
-      continue;
+  // Recursively hoist all the producers of `index`
+  bool changed = false; // if any of the inputs is replaced by an existing val
+  std::vector<Val*> inputs;
+  for (auto input : def->inputs()) {
+    inputs.emplace_back(hoistIndex(input, loops, my_level));
+    if (inputs.back() == nullptr) {
+      return nullptr;
     }
-    // Initial index variables can have some additions such as magic
-    // zero and "1" when used in producer indexing for double buffered
-    // tensors. Thus, the initial variables themselves may be
-    // different, and its components need to be examined. An easy way
-    // is to flatten them to strings as follows.
-    auto lhs_str = loop_index_vals_.at(i)->toInlineString();
-    auto rhs_str = other.loop_index_vals_.at(i)->toInlineString();
-    if (lhs_str == rhs_str) {
-      continue;
+    if (inputs.back() != input) {
+      changed = true;
     }
+  }
 
-    return false;
+  // If any of the inputs is replaced, then create a new expression whose inputs
+  // are replaced with hoisted input
+  if (changed) {
+    index = IrBuilder::newScalar(*index->getDataType());
+    TORCH_INTERNAL_ASSERT(def->outputs().size() == 1);
+    auto create_fn = def->newObjectFunc();
+    create_fn(index->container(), inputs, {index}, def->attributes());
```

Comment:
I am so happy that now I can create an arbitrary expression without knowing what type it is.

Reply:
This is really cool!